File: packages/database/src/repositories/runs.ts
import { assertRevision, nextRevision } from "../revision";

export type RunStatus = "pending" | "running" | "completed" | "partial" | "failed" | "cancelled";
export interface AutomationRun {
  id: string;
  ownerId: string;
  status: RunStatus;
  revision: number;
  correlationId: string;
  idempotencyKey: string;
}

const transitions: Record<RunStatus, readonly RunStatus[]> = {
  pending: ["running", "cancelled"],
  running: ["completed", "partial", "failed", "cancelled"],
  completed: [],
  partial: [],
  failed: [],
  cancelled: []
};

export function transitionRun(
  run: AutomationRun,
  status: RunStatus,
  expectedRevision: number
): AutomationRun {
  assertRevision(run.revision, expectedRevision);
  if (!transitions[run.status].includes(status))
    throw new Error(`Invalid run transition ${run.status} -> ${status}`);
  return { ...run, status, revision: nextRevision(run.revision) };
}

File: apps/worker/src/career_worker/repositories/runs.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime


@dataclass(slots=True)
class Run:
    id: str
    owner_id: str
    status: str = "pending"
    steps: dict[str, str] = field(default_factory=dict)
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    def transition(self, status: str) -> None:
        allowed = {
            "pending": {"running", "cancelled"},
            "running": {"completed", "partial", "failed", "cancelled"},
        }
        if status not in allowed.get(self.status, set()):
            raise ValueError(f"invalid run transition: {self.status} -> {status}")
        self.status = status
        self.updated_at = datetime.now(UTC)
